src/predictor/cpu_predictor.cc

```diff
@@ -99,51 +120,49 @@ class AdapterView {
   bst_row_t const static base_rowid = 0;  // NOLINT
 };
 
-template <typename DataView>
-void PredictBatchKernel(DataView batch, std::vector<bst_float> *out_preds,
+template <typename DataView, size_t block_of_rows_size>
+void PredictBatchByBlockOfRowsKernel(DataView batch, std::vector<bst_float> *out_preds,
                         gbm::GBTreeModel const &model, int32_t tree_begin,
                         int32_t tree_end,
                         std::vector<RegTree::FVec> *p_thread_temp) {
   auto& thread_temp = *p_thread_temp;
+  const std::vector<std::unique_ptr<RegTree>> &trees = model.trees;
   int32_t const num_group = model.learner_model_param->num_output_group;
 
   std::vector<bst_float> &preds = *out_preds;
   CHECK_EQ(model.param.size_leaf_vector, 0)
       << "size_leaf_vector is enforced to 0 so far";
   // parallel over local batch
   const auto nsize = static_cast<bst_omp_uint>(batch.Size());
-  auto constexpr kUnroll = DataView::kUnroll;
-  const bst_omp_uint rest = nsize % kUnroll;
-  if (nsize >= kUnroll) {
-#pragma omp parallel for schedule(static)
-    for (bst_omp_uint i = 0; i < nsize - rest; i += kUnroll) {
-      const int tid = omp_get_thread_num();
-      RegTree::FVec &feats = thread_temp[tid];
-      int64_t ridx[kUnroll];
-      SparsePage::Inst inst[kUnroll];
-      for (size_t k = 0; k < kUnroll; ++k) {
-        ridx[k] = static_cast<int64_t>(batch.base_rowid + i + k);
-      }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        inst[k] = batch[i + k];
-      }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        for (int gid = 0; gid < num_group; ++gid) {
-          const size_t offset = ridx[k] * num_group + gid;
-          preds[offset] += PredValue(inst[k], model.trees, model.tree_info, gid,
-                                     &feats, tree_begin, tree_end);
+
+  const size_t n_blocks = omp_get_max_threads();
+  const size_t block_size = nsize / n_blocks;
+  #pragma omp parallel
```

Comment:
General question - why do we use ```omp parallel```, but not ```omp parallel for``` by blocks of rows what you process below?
Because in case of ```omp parallel``` you need have 2 levels of blocking: threads and rows, for ```omp parallel for``` only by rows

Reply:
changed to `omp parallel for`

Reply:
blocking was simplified. Thanks